### PR TITLE
chore: Test the main contract

### LINKED ISSUE #98: Add tests for QGB itself

### PR DESCRIPTION
**Pull request by evan-forbes**
This PR starts the tests for the main contracts. 

While this is a decent start, I'm not sure we can close #98 as this is still far from appropriate coverage percentage imho. I didn't originally set out to close #98, only do complete enough to help me debug the relayer. We still need to test expected failures, nil signatures, and a few other edge cases.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "ethereum/solidity/lib/ds-test"]
 	path = ethereum/solidity/lib/ds-test
 	url = https://github.com/dapphub/ds-test
+[submodule "ethereum/solidity/lib/openzeppelin-contracts"]
```
Comment (SweeXordious) [inline .gitmodules:4]: I wonder how were we able to run tests before without having this set here :D

**Review by SweeXordious — APPROVED** [review]
LGTM.
Thanks for adding the checkpoint test :D

File: ethereum/solidity/src/test/QuantumGravityBridge.t.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    // testAddr  = "0x9c2B12b5a07FC6D719Ed7646e5041A7E85758329"
+    // testPriv  = "64a1d6f0e760a8d62b4afdde4096f16f51b401eaaecc915740f71770ea76a8ad"
+    // testAddr2 = "0xe650B084f05C6194f6e552e3b9f08718Bc8a9d56"
+    // testPriv2 = "6e8bdfa979ab645b41c4d17cb1329b2a44684c82b61b1b060ea9b6e1c927a4f4"
+
+    bytes32 constant VALIDATOR_SET_HASH_DOMAIN_SEPARATOR =
+        0x636865636b706f696e7400000000000000000000000000000000000000000000;
+
+    bytes32 constant DATA_ROOT_TUPLE_ROOT_DOMAIN_SEPARATOR =
+        0x7472616e73616374696f6e426174636800000000000000000000000000000000;
```
Comment (liamsi) [inline ethereum/solidity/src/test/QuantumGravityBridge.t.sol:23]: Are these domain separators chosen arbitrarily for this test?

**Review by liamsi — COMMENTED** [review]
I'm not a solidity expert but this LGTM. Can we reject PRs without unit tests in the future? They should be part of any feature introduced.

File: ethereum/solidity/src/test/QuantumGravityBridge.t.sol (hunk 2 above)
Comment (adlerjohn) [reply ethereum/solidity/src/test/QuantumGravityBridge.t.sol:23]: No, they're in the contract

https://github.com/celestiaorg/quantum-gravity-bridge/blob/cc0560976538fc3bde396e9d1fa4c7a9b1aefc75/ethereum/solidity/src/QuantumGravityBridge.sol#L43-L49

File: .gitmodules (hunk 1 above)
Comment (adlerjohn) [reply .gitmodules:4]: Ah, so `forge build` doesn't actually error if it can't find an import... `forge test` does, but given that the QGB contract itself was not tested prior to this PR, then of course there would be no error.

This is of course one of the many examples to Solidity tooling being extremely immature. In the future we should examine using [Sway](https://github.com/FuelLabs/sway) and relaying directly to a Fuel execution layer.

File: ethereum/solidity/src/test/QuantumGravityBridge.t.sol [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import "../lib/openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "../DataRootTuple.sol";
+import "../QuantumGravityBridge.sol";
+
+import "ds-test/test.sol";
+
+contract RelayerTest is DSTest {
+
+    // private keys used for hardcoded signatures signatures
```
Comment (adlerjohn) [inline ethereum/solidity/src/test/QuantumGravityBridge.t.sol:13]:
```suggestion
contract RelayerTest is DSTest {
    // private keys used for hardcoded signatures
```

File: ethereum/solidity/src/test/QuantumGravityBridge.t.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        uint256 _nonce,
+        uint256 _powerThreshold,
+        bytes32 _validatorSetHash
+    ) private pure returns (bytes32) {
+        bytes32 c = keccak256(
+            abi.encode(_bridge_id, VALIDATOR_SET_HASH_DOMAIN_SEPARATOR, _nonce, _powerThreshold, _validatorSetHash)
+        );
+
+        return c;
+    }
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply ethereum/solidity/src/test/QuantumGravityBridge.t.sol:95]: I tried, but couldn't get the mocks to work. per the docs

> Private functions and state variables are only visible for the contract they are defined in and not in derived contracts.

does this mean that we can't use them in a mock contract? I could still be doing something wrong

[the remark this answers is not in the document]
Comment (evan-forbes) [reply ethereum/solidity/src/test/QuantumGravityBridge.t.sol:95]: this won't compile

```solidity
contract ExposedQuantumGravityBridge is QuantumGravityBridge {
    function _computeValidatorSetHash(Validator[] memory _validators) public pure returns (bytes32) {
        return computeValidatorSetHash(_validators);
    }

    function _domainSeparateValidatorSetHash(
        bytes32 _bridge_id,
        uint256 _nonce,
        uint256 _powerThreshold,
        bytes32 _validatorSetHash
    ) public pure returns (bytes32) {
        return domainSeparateValidatorSetHash(_bridge_id, _nonce, _powerThreshold, _validatorSetHash);
    }
}
```

**Review by adlerjohn — APPROVED** [review]
